dhall-docs/src/Dhall/Docs/CodeRenderer.hs

```diff
@@ -0,0 +1,149 @@
+                HTTP -> "http"
+                HTTPS -> "https"
+
+            path_ = fileAsText path
+
+            query_ = case query of
+                Nothing -> ""
+                Just d -> "?" <> d
+
+            -- we don't include the headers here since we treat links to open a file
+            -- in another tab
+            href = scheme_ <> "://" <> authority <> path_ <> query_
+
+        Local Here file -> a_ [href_ href] . toHtml
+          where
+            href = "." <> fileAsText file <> ".html"
+
+        Local Parent file -> a_ [href_ href] . toHtml
+          where
+            href = ".." <> fileAsText file <> ".html"
+
+        _ -> toHtml
+
+-- | Given a Text and the parsed `Expr Src Import` from it, this will render the
+--   the source code on HTML with jump-to-definition on URL imports
+exprSrcToHtml :: Text -> Expr Src Import -> Html ()
+exprSrcToHtml contents expr = pre_ $ go (1, 1) (Data.Text.lines contents) imports
+  where
+    imports = getImports expr
+    sourceLine = SourcePos.unPos . SourcePos.sourceLine
+    sourceColumn = SourcePos.unPos . SourcePos.sourceColumn
+
+    -- we keep the current line, column and consumed text as part of function argument
+    go :: (Int, Int) -> [Text] -> [(Src, Import)] -> Html ()
+    go _ textLines [] = mapM_ (\t -> toHtml t >> br_ []) textLines
+    go (currLine, currCol) currentLines ((Src {..}, import_) : rest) = do
+        prefix
+        renderImport import_ srcText
+        if Data.Text.null suffixCols then br_ [] else return ()
+        go nextPosition suffix rest
+      where
+        importStartLine = sourceLine srcStart
+        importEndLine = sourceLine srcEnd
+
+        importStartCol = sourceColumn srcStart
+        importEndCol = sourceColumn srcEnd
+
+        (prefixLines, restLines) = splitAt (importStartLine - currLine) currentLines
+        (importLines, suffixLines) = splitAt (importStartLine - importStartLine + 1) restLines
+
+        -- calls to `head` and `last` here should never fail since `importLines`
+        -- have at least one element
+        (firstImportLine, lastImportLine) = (head importLines, last importLines)
+        prefixCols = Data.Text.take (importStartCol - currCol) firstImportLine
+        suffixCols = Data.Text.drop (importEndCol - currCol) lastImportLine
+
+        prefix = mapM_ (\t -> toHtml t >> br_ []) prefixLines >> toHtml prefixCols
+        suffix = if Data.Text.null suffixCols then suffixLines else suffixCols : suffixLines
+        nextPosition =
+            if Data.Text.null suffixCols then
+                (importEndLine + 1, 1)
+            else (importEndLine, importEndCol)
+
+-- | Internal utility to differentiate if a Dhall expr is a type annotation
+--   or the whole file
+data ExprType = TypeAnnotation | AssertionExample
+
+exprVoidToHtml :: Dhall.Pretty.CharacterSet -> ExprType -> Expr Void Import -> Html ()
+exprVoidToHtml characterSet exprType expr = exprSrcToHtml formattedFile expr'
```

Comment:
This one needs haddocks. I currently don't understand why there's both `exprSrcToHtml` and `exprVoidToHtml`.

Reply:
I added haddocks for that function with a clear explanation and documented the use-case for each one.